compiler/src/iree/compiler/Dialect/Flow/TransformExtensions/FlowExtensions.cpp

```diff
@@ -446,5 +442,234 @@ transform_dialect::ForeachThreadToFlowDispatchWorkgroupsOp::applyToOne(
+
+/// Compute the dynamic dims of the given value and add them to the vector.
+static void appendDynamicDims(OpBuilder &b, Location loc,
+                              SmallVector<Value> &argumentDims, Value tensor) {
+  auto tensorType = tensor.getType().cast<RankedTensorType>();
+
+  // Fast-path for if the value comes from ops that support our dynamic
+  // shape interfaces. Otherwise we have to insert tensor.dim ops.
+  auto availableDims = IREE::Util::findDynamicDims(tensor);
+  if (availableDims.hasValue()) {
+    argumentDims.append(availableDims->begin(), availableDims->end());
+    assert(tensorType.getNumDynamicDims() == availableDims->size() &&
+           "not enough dynamic dims found");
+    return;
+  }
+
+  for (auto dim : llvm::enumerate(tensorType.getShape())) {
+    if (dim.value() != ShapedType::kDynamicSize) continue;
+    argumentDims.push_back(
+        b.createOrFold<tensor::DimOp>(loc, tensor, dim.index()));
+  }
+}
+
+/// Follow the reverse SSA use-def chain of the given value (always taking the
+/// tied operand) and return the first value outside of `containingOp`.
+static Optional<Value> findFirstTiedValueOutsideOfContainingOp(
+    Operation *containingOp, Value value) {
+  // Check if `v` is defined outside of `containing_op`.
+  auto isOutside = [&](Value v) {
+    if (v.isa<OpResult>() && !containingOp->isAncestor(v.getDefiningOp()))
+      return true;
+    if (v.isa<BlockArgument>()) {
+      assert(
+          v.cast<BlockArgument>().getOwner()->getParentOp()->isProperAncestor(
+              containingOp) &&
+          "expected that bbArgs belong to a parent of the container");
+      return true;
+    }
+    return false;
+  };
+
+  while (!isOutside(value)) {
+    auto tiedOpInterface = value.getDefiningOp<IREE::Util::TiedOpInterface>();
+    if (!tiedOpInterface)
+      // Reached an op that does not implement the interface.
+      return llvm::None;
+    value = tiedOpInterface.getTiedResultOperand(value);
+    if (!value)
+      // Nothing is tied here.
+      return llvm::None;
+  }
+
+  return value;
+}
+
+/// Rewrite the DispatchRegionOp into a DispatchWorkgroupsOp. The
+/// DispatchRegionOp is not isolated from above and may capture any SSA value
+/// that is in scope. The generated DispatchWorkgroupsOp captures all SSA values
+/// explicitly and makes them available inside the region via block arguments.
+static FailureOr<Flow::DispatchWorkgroupsOp>
+rewriteFlowDispatchRegionToFlowDispatchWorkgroups(
+    Flow::DispatchRegionOp regionOp, RewriterBase &rewriter) {
+  // Only ops with a single block are supported.
+  Region &region = regionOp.getBody();
+  if (!region.hasOneBlock()) return failure();
+  Block &body = region.front();
+
+  // All results shapes must be reifiable.
+  auto terminator = cast<Flow::ReturnOp>(body.getTerminator());
+  unsigned numResults = terminator->getNumOperands();
+  if (!llvm::all_of(terminator->getOperands(), [](Value v) {
```

Comment:
I dont think we need this here. The dynamic dims of the result are better captured during formation of the `flow.dispatch.region` op itself. Here we just move the dynamic dims over?